### PR TITLE
Fix #537; Ability to specify number of checkpoints that are kept in c…

### LINKED ISSUE #537: Ability to specify number of checkpoints that are kept in checkpoint dir
Pass a [CheckpointSaverHook](https://www.tensorflow.org/api_docs/python/tf/train/CheckpointSaverHook) instance with a custom `Saver` to `MonitoredTrainingSession` and expose the saver's `max_to_keep` property as a `FLAGS` variable.
As discussed in https://github.com/mozilla/DeepSpeech/issues/534 .

### PR DESCRIPTION
**Pull request by andi4191**
…heckpoint dir

@tilmankamp, @kdavis-mozilla : Request you to have a look at this.

### CODE REVIEW

**Comment by andi4191** [comment]
@kdavis-mozilla, @tilmankamp : I have tested this fix with configuring the max_to_keep flags parameter.

File: DeepSpeech.py [hunk 1]
```diff
@@ -1444,6 +1445,11 @@ def end(self, session):
     if FLAGS.summary_secs > 0:
         hooks.append(tf.train.SummarySaverHook(save_secs=FLAGS.summary_secs, output_dir=FLAGS.summary_dir, summary_op=merge_all_summaries_op))
 
+    # Hook wih number of checkpoint files to save in checkpoint_dir
+    if FLAGS.max_to_keep > 0:
+	saver_var = tf.train.Saver(max_to_keep=FLAGS.max_to_keep)
+        hooks.append(tf.train.CheckpointSaverHook(checkpoint_dir=FLAGS.checkpoint_dir, save_secs=FLAGS.checkpoint_secs, saver=saver_var))
```
Comment (tilmankamp) [inline DeepSpeech.py]: This collides with the construction of `MonitoredTrainingSession`. I think the checkpoint parameters should be removed from `MonitoredTrainingSession`.

File: DeepSpeech.py [hunk 2]
```diff
@@ -96,6 +96,7 @@
 
 tf.app.flags.DEFINE_string  ('checkpoint_dir',   '',          'directory in which checkpoints are stored - defaults to directory "deepspeech/checkpoints" within user\'s data home specified by the XDG Base Directory Specification')
 tf.app.flags.DEFINE_integer ('checkpoint_secs',  600,         'checkpoint saving interval in seconds')
+tf.app.flags.DEFINE_integer ('max_to_keep',      10,           'number of checkpoint files to keep - default value is 5')
```
Comment (tilmankamp) [inline DeepSpeech.py]: I'd use the default number from TensorFlow here.

File: DeepSpeech.py (hunk 1 above)
Comment (andi4191) [inline DeepSpeech.py]: @tilmankamp: I am not sure if I get it. If we remove the CheckpointSaverHook from MonitoredTrainingSession. How would it limit the number of checkpoint files? MonitoredTrainingSession can have multiple hooks passed as an argument.

Comment (tilmankamp) [inline DeepSpeech.py]: I was talking about `FLAGS.checkpoint_dir`and `FLAGS.checkpoint_secs` in
```
         with tf.train.MonitoredTrainingSession(master='' if server is None else server.target,
                                                is_chief=is_chief,
                                                hooks=hooks,
                                                checkpoint_dir=FLAGS.checkpoint_dir,
                                                save_checkpoint_secs=FLAGS.checkpoint_secs,
                                                config=session_config) as session:
```

Comment (andi4191) [inline DeepSpeech.py]: Oh! checkpoint_dir and either one of save_secs or save_steps are mandatory arguments.
Should I specify another flags for paths and save_secs in order to remove the coupling between the two sets of arguments?

File: DeepSpeech.py (hunk 2 above)
Comment (andi4191) [inline DeepSpeech.py]: ok will change it to 5

File: DeepSpeech.py (hunk 1 above)
Comment (tilmankamp) [inline DeepSpeech.py]: Oh - wasn't aware of that. Is your hook then overruling this settings?

Comment (andi4191) [inline DeepSpeech.py]: I have updated the tf.train.CheckpointSaverHook with the checkpoint_dir and save_secs and removed the same parameters from the MonitoredTrainingSession. The result is the same. The coupling between the two set of arguments has been removed now and it works as expected.

File: DeepSpeech.py [hunk 3]
```diff
@@ -1444,15 +1445,18 @@ def end(self, session):
     if FLAGS.summary_secs > 0:
         hooks.append(tf.train.SummarySaverHook(save_secs=FLAGS.summary_secs, output_dir=FLAGS.summary_dir, summary_op=merge_all_summaries_op))
 
+    # Hook wih number of checkpoint files to save in checkpoint_dir
+    if FLAGS.max_to_keep > 0:
+	saver_var = tf.train.Saver(max_to_keep=FLAGS.max_to_keep)
```
Comment (tilmankamp) [inline DeepSpeech.py]: `saver_var` is a strange name - please change it so something else.
Also: indenting

**Pull request merged**